File: lang/python/test_id_abi.py
"""Tests for the ID/IDREF C ABI surface (cx_id_lookup / cx_resolve_ref /
cx_node_id) per ADR 0003 / Phase 7.65. The Document-level resolve_id /
elements_by_id surface is already covered by test_identity.py; these
tests exercise the stateless string-in/string-out C ABI variants.
"""
import json
from cxlib.cx import id_lookup, resolve_ref, node_id


DOC = (
    "[users\n"
    "  [user #u-1 name=alice]\n"
    "  [user #u-2 name=bob]\n"
    "  [reviewer assigned-to=@u-1]\n"
    "]"
)


def test_id_lookup_returns_ast_json_for_declared_id():
    out = id_lookup(DOC, "u-1")
    assert out is not None
    payload = json.loads(out)
    assert payload["type"] == "Element"
    assert payload["name"] == "user"
    assert payload["id"] == "u-1"


def test_id_lookup_missing_returns_none():
    assert id_lookup(DOC, "does-not-exist") is None


def test_resolve_ref_matches_id_lookup():
    a = id_lookup(DOC, "u-2")
    b = resolve_ref(DOC, "u-2")
    assert a == b


def test_node_id_returns_id_for_id_bearing_element():
    assert node_id(DOC, "//user") == "u-1"


def test_node_id_empty_for_non_id_bearing_element():
    assert node_id(DOC, "//reviewer") is None


def main():
    import sys, traceback
    tests = [g for n, g in sorted(globals().items()) if n.startswith("test_") and callable(g)]
    passed = failed = 0
    for t in tests:
        try:
            t()
            print(f"ok  {t.__name__}")
            passed += 1
        except Exception:
            traceback.print_exc()
            print(f"FAIL {t.__name__}")
            failed += 1
    print(f"\n{passed}/{passed + failed} passed")
    sys.exit(1 if failed else 0)


if __name__ == "__main__":
    main()

File: lang/ruby/test_id_abi.rb
# frozen_string_literal: true
#
# ID/IDREF C ABI tests for the Ruby CX binding (ADR 0003 / Phase 7.65).
# Mirrors lang/python/test_id_abi.py and lang/go/cxlib/id_abi_test.go.
# Run: ruby lang/ruby/test_id_abi.rb
#
require 'json'
require_relative 'cxlib/lib/cxlib'

$passed = 0
$failed = 0

def run_test(name)
  yield
  $passed += 1
  print "  PASS  #{name}\n"
rescue => e
  $failed += 1
  print "  FAIL  #{name}\n"
  print "    #{e.class}: #{e.message}\n"
  print "    #{e.backtrace.first(2).join("\n    ")}\n"
end

def assert_eq(expected, actual, msg = nil)
  raise "expected #{expected.inspect}, got #{actual.inspect}#{" (#{msg})" if msg}" \
    if expected != actual
end

def assert_true(cond, msg = nil)
  raise "expected true#{" (#{msg})" if msg}" unless cond
end

DOC = <<~CX
  [users
    [user #u-1 name=alice]
    [user #u-2 name=bob]
    [reviewer assigned-to=@u-1]
  ]
CX

run_test('id_lookup_happy_path') do
  out = CXLib.id_lookup(DOC, 'u-1')
  assert_true !out.empty?, 'expected non-empty AST-JSON for #u-1'
  obj = JSON.parse(out)
  assert_eq 'Element', obj['type']
  assert_eq 'user',    obj['name']
  assert_eq 'u-1',     obj['id']
end

run_test('id_lookup_missing_returns_empty') do
  out = CXLib.id_lookup(DOC, 'does-not-exist')
  assert_eq '', out
end

run_test('resolve_ref_equals_id_lookup') do
  a = CXLib.id_lookup(DOC, 'u-2')
  b = CXLib.resolve_ref(DOC, 'u-2')
  assert_true !a.empty?, 'id_lookup u-2 should return non-empty'
  assert_eq a, b
end

run_test('node_id_at_cxpath') do
  assert_eq 'u-1', CXLib.node_id(DOC, '//user')
  assert_eq '',    CXLib.node_id(DOC, '//reviewer')
end

print "\n#{$passed} passed, #{$failed} failed\n"
exit($failed.zero? ? 0 : 1)
